java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java

```diff
@@ -68,31 +77,64 @@ public void leaveFile(JavaFileScannerContext context) {
   @Override
   public void visitNode(Tree tree) {
     TryStatementTree tryStatementTree = (TryStatementTree) tree;
+    List<CatchTree> catchGenericException = new ArrayList<>();
+    boolean interruptCaught = false;
+
     withinInterruptingFinally.addFirst(isFinallyInterrupting(tryStatementTree.finallyBlock()));
     for (CatchTree catchTree : tryStatementTree.catches()) {
-      Optional<Type> interruptType = findInterruptingType(catchTree.parameter());
-      if(interruptType.isPresent()) {
-        BlockVisitor blockVisitor = new BlockVisitor(catchTree.parameter().symbol());
-        catchTree.block().accept(blockVisitor);
-        if(!blockVisitor.threadInterrupted && !isWithinInterruptingFinally()) {
-          reportIssue(catchTree.parameter(), "Either re-interrupt this method" +
-            " or rethrow the \""+interruptType.get().name()+"\".");
+      VariableTree catchParameter = catchTree.parameter();
+      Optional<Type> interruptType = getIfCatchMatch(catchParameter, INTERRUPTING_TYPE_PREDICATE);
+      if (interruptType.isPresent()) {
+        interruptCaught = true;
+        if (handleInCorrectlyInterruption(catchTree)) {
+          reportIssue(catchParameter, String.format(MESSAGE, interruptType.get().name()));
         }
```

Comment:
It seems to me we should stop iterating, once we found appropriate Exception handler (InterruptedException, Exception or Throwable) because only one catch block will be executed once exception is caught.

Reply:
It makes a lot of sense, doing this will clean the flow of the logic, and consistently support all the examples.